[Rocprofiler-system]: Add Documentation for communication API tracing (#2478)
ROCm/rocm-systems: projects/rocprofiler-systems/docs/how-to/communication-runtime-profiling.rst

.. meta::
   :description: ROCm Systems Profiler communication runtime profiling documentation
   :keywords: rocprof-sys, rocprofiler-systems, ROCm, MPI, RCCL, UCX, communication, profiler, tracking, distributed, AMD

****************************************************
Communication Runtime Profiling
****************************************************

> `ROCm Systems Profiler <https://github.com/ROCm/rocm-systems/tree/develop/projects/rocprofiler-systems>`_ can be used to profile several widely used communication runtimes and libraries, including MPI, RCCL, and UCX.

Reviewer: ```suggestion
`ROCm Systems Profiler <https://github.com/ROCm/rocm-systems/tree/develop/projects/rocprofiler-systems>`_ profiles several widely used communication runtimes and libraries, including MPI, RCCL, and UCX.
```

Slightly more concise.

* **RCCL (ROCm Communication Collectives Library)**: AMD's GPU-aware collective communication library, optimized for multi-GPU communication within and across nodes. RCCL is designed to work seamlessly with ROCm and provides highly optimized implementations of collective operations like AllReduce, AllGather, and Broadcast.

**Low-Level Communication Frameworks**

* **UCX (Unified Communication X)**: A high-performance communication framework that provides low-level abstractions for RDMA, shared memory, and other transport mechanisms. UCX is often used as a backend for higher-level libraries like MPI and RCCL, providing efficient point-to-point communication, RMA (Remote Memory Access) operations, and active messages.

.. note::

   **Automatic Detection and Default Behavior:**

   * **MPI** (``ROCPROFSYS_USE_MPIP``): Enabled by default (``true``). When using binary instrumentation, ROCm Systems Profiler automatically detects MPI symbols in the target application and enables MPI support.
   * **UCX** (``ROCPROFSYS_USE_UCX``): Enabled by default (``true``). Automatically intercepts UCX functions if the UCX library is loaded by the application.

>    * **RCCL** (``ROCPROFSYS_USE_RCCLP``): Disabled by default (``false``). Must be explicitly enabled to trace RCCL operations.

Reviewer: ```suggestion
   * **MPI** (``ROCPROFSYS_USE_MPIP``): Enabled by default (``ON``). When using binary instrumentation, ROCm Systems Profiler automatically detects MPI symbols in the target application and enables MPI support.
   * **UCX** (``ROCPROFSYS_USE_UCX``): Enabled by default (``ON``). Automatically intercepts UCX functions if the UCX library is loaded by the application.
   * **RCCL** (``ROCPROFSYS_USE_RCCLP``): Disabled by default (``OFF``). Must be explicitly enabled to trace RCCL operations.
```
Suggest using "ON" / "OFF", rather than "true" / "false", because that seems to be what is used most throughout the page.

   These settings can be controlled at runtime using their respective environment variables to enable or disable tracing as needed.

Profiling MPI
=============

> MPI support is enabled through the ``ROCPROFSYS_USE_MPIP`` configuration setting, which is **enabled by default**. ROCm Systems Profiler can have full (``ROCPROFSYS_USE_MPI=ON``) or partial (``ROCPROFSYS_USE_MPI_HEADERS=ON``) MPI support. By default, ROCm Systems Profiler uses partial MPI support with the OpenMPI headers. For detailed information on building rocprofiler-systems with MPI support, see the :doc:`installation guide <../install/install>`.

Reviewer: Make sure it is clear that "ROCPROFSYS_USE_MPI" and "ROCPROFSYS_USE_MPI_HEADERS" are build-time settings, rather than run-time.
Maybe, "ROCm Systems Profiler can be built with ..."

   export ROCPROFSYS_TRACE=ON
   export ROCPROFSYS_PROFILE=ON

   # To explicitly disable MPI profiling if needed:
   export ROCPROFSYS_USE_MPIP=OFF

When MPI support is enabled, rocprofiler-systems automatically intercepts MPI function calls using GOTCHA wrappers, allowing you to trace MPI communication patterns and timing.

Usage with MPI Applications
----------------------------

When profiling MPI applications, use ``rocprof-sys-sample`` instead of ``rocprof-sys-instrument`` with runtime instrumentation to avoid compatibility issues with MPI process launching:

.. code-block:: shell

   # Recommended: Using rocprof-sys-sample
   mpirun -n 4 rocprof-sys-sample -- ./my_mpi_app

   # Alternative: Binary rewrite approach
   rocprof-sys-instrument -o my_mpi_app.inst -- ./my_mpi_app
   mpirun -n 4 rocprof-sys-run -- ./my_mpi_app.inst

.. note::

   Runtime instrumentation (``rocprof-sys-instrument`` without ``-o``) requires a fork and ``ptrace``, which is generally incompatible with how MPI applications spawn processes, particularly with OpenMPI.

MPI Profiling Output
--------------------

When MPI profiling is enabled, ROCm Systems Profiler generates:

> * **Perfetto traces**: Visualize MPI operations on a timeline, showing communication patterns, operation durations, and concurrency

Reviewer: Do we want to mention rocpd output too?

* **Timemory profiles**: Statistical summaries of MPI function call counts, total time, and performance metrics
* **Communication data**: Track message sizes, communication volumes, and data movement patterns for point-to-point and collective operations

The traces include detailed information about:

* MPI ranks and communicators
* Message sizes and datatypes
* Source and destination ranks (for point-to-point operations)
* Root ranks (for collective operations)
* Tags for message matching

ROCm Systems Profiler provides automatic output labeling based on MPI rank IDs:

* When full MPI support is enabled (``ROCPROFSYS_USE_MPI=ON``), output files are labeled with the ``MPI_COMM_WORLD`` rank ID
* The ``ROCPROFSYS_USE_PID`` setting controls whether process IDs or MPI rank IDs are used for output labeling

For detailed information on building rocprofiler-systems with MPI support, see the :doc:`installation guide <../install/install>`.

Profiling RCCL
==============

> RCCL (ROCm Communication Collectives Library) profiling provides insights into GPU-to-GPU communication patterns and collective operation performance. 

Reviewer: ```suggestion
RCCL profiling provides insights into GPU-to-GPU communication patterns and collective operation performance.
```

nit: since you have already defined "RCCL" above, you don't have to expand it here too.
formatting: removed trailing whitespace.

.. important::

   Unlike MPI and UCX, RCCL profiling is **disabled by default** and must be explicitly enabled using ``ROCPROFSYS_USE_RCCLP=ON``.

When enabled, rocprofiler-systems captures:

* RCCL API calls (ncclAllReduce, ncclBroadcast, ncclReduce, etc.)
* Communication data volumes and patterns
* Timing information for collective operations

Configuration
-------------

To enable RCCL tracing and profiling:

.. code-block:: shell

   export ROCPROFSYS_USE_RCCLP=ON
   export ROCPROFSYS_TRACE=ON
   export ROCPROFSYS_PROFILE=ON
   export ROCPROFSYS_ROCM_DOMAINS=hip_runtime_api,kernel_dispatch,memory_copy

RCCL Profiling Output
-------------

The image below shows an example of a Perfetto trace with RCCL communication data and API tracing enabled:

.. image:: ../data/rccl-comm-recv.png
   :alt: Perfetto tracks with RCCL Communication Data and API tracing

In the Perfetto trace, you can observe:

* RCCL collective operations on dedicated tracks
* Communication volume and direction
* Overlap between computation and communication
* Synchronization points and barriers

.. note::

>    There is a known issue which causes the application to exit with an error. However, the trace data can still be found in the output directory. This issue is being tracked internally.

Reviewer: Note, this issue is fixed in 7.12, so we can reword this.

```suggestion
In ROCm versions prior to 7.12, there is a known issue which causes the application to exit with an error. However, the trace data can still be found in the output directory. This issue has been resolved in ROCm 7.12 and later versions.
```

Profiling UCX
=============

> UCX (Unified Communication X) is a low-level communication framework that provides the foundation for efficient data movement in high-performance computing applications. UCX profiling enables detailed analysis of low-level communication primitives, RDMA operations, and transport-layer behavior.

Reviewer: ```suggestion
UCX is a low-level communication framework that provides the foundation for efficient data movement in high-performance computing applications. UCX profiling enables detailed analysis of low-level communication primitives, RDMA operations, and transport-layer behavior.
```

nit: Since you define UCX above, you don't need to expand it here too.

> UCX profiling is **enabled by default** (``ROCPROFSYS_USE_UCX=ON``). When an application uses UCX—either directly or indirectly through higher-level libraries like MPI or RCCL—rocprofiler-systems automatically intercepts and traces UCX function calls.

Reviewer: ```suggestion
UCX profiling is **enabled by default** (``ROCPROFSYS_USE_UCX=ON``). When an application uses UCX — either directly or indirectly through higher-level libraries like MPI or RCCL — rocprofiler-systems automatically intercepts and traces UCX function calls.
```
I am not sure if RST requires formatting like that, but just looking at the text, it looks like there should be spaces on either side of that `-`. Ignore this comment if it is a RST requirement.

Configuration
-------------

> Since UCX profiling is enabled by default, you typically don't need to explicitly enable it:

Reviewer: ```suggestion
Since UCX profiling is enabled by default, you typically don't need to explicitly enable it.  However, if you need to disable UCX tracing, you can do so with the following configuration settings.
```

**Active Messages**

Active messages provide low-latency communication with handler execution on the receiver:

* ``ucp_am_send_nbx`` - Non-blocking active message send
* ``ucp_am_recv_data_nbx`` - Non-blocking active message receive

**Atomic Operations**

UCX provides various atomic operations for lock-free algorithms and synchronization:

* ``ucp_atomic_add32``, ``ucp_atomic_add64`` - Atomic addition
* ``ucp_atomic_fadd32``, ``ucp_atomic_fadd64`` - Fetch-and-add
* ``ucp_atomic_swap32``, ``ucp_atomic_swap64`` - Atomic swap
* ``ucp_atomic_cswap32``, ``ucp_atomic_cswap64`` - Compare-and-swap

**Stream Operations**

Stream operations provide ordered, connection-oriented communication:

* ``ucp_stream_send_nbx`` - Non-blocking stream send
* ``ucp_stream_recv_nbx`` - Non-blocking stream receive

Usage with UCX Applications
----------------------------

UCX profiling works transparently with applications that use UCX directly or indirectly through higher-level libraries:

.. code-block:: shell

   # Direct UCX application
   rocprof-sys-run -- ./my_ucx_app

   # MPI application using UCX as transport
   export ROCPROFSYS_USE_MPIP=ON
   export ROCPROFSYS_USE_UCX=ON

>    mpirun -n 4 rocprof-sys-sample -- ./my_mpi_ucx_app

Reviewer: ```suggestion
   # Example 1: Direct UCX application
   rocprof-sys-sample -- ./my_ucx_app

   # Example 2: MPI application using UCX as transport
   export ROCPROFSYS_USE_MPIP=ON
   export ROCPROFSYS_USE_UCX=ON
   mpirun -n 4 rocprof-sys-sample -- ./my_mpi_ucx_app
```

Adjusted the comments to make is more explicit that these were two different examples.
Also, used `rocprof-sys-sample` for both so we don't accidently imply that the uses has to use "sample" in one case and "run" in the other.

.. note::

>    For MPI applications, the presence of UCX libraries alone does not ensure UCX is used at runtime. When MPI is launched with the UCX PML ( ``-mca pml ucx`` ), initialization may fail due to UCX version or transport capability mismatches, causing MPI to fall back to an alternative (non-UCX) communication path.

Reviewer: This is a great note. Can the user also verify that UCX is used with some sort of verbose logging? I have read that `--mca pml_base_verbose #` can be used?

UCX Profiling Output
---------------------

When UCX profiling is enabled, rocprofiler-systems generates:

* **Perfetto traces**: Visualize UCX operations on a timeline, showing communication patterns, operation durations, and concurrency
* **Timemory profiles**: Statistical summaries of UCX function call counts, total time, and performance metrics

> * **Communication data**: Track message sizes, communication volumes, and data movement patterns

Reviewer: Do we want to mention rocpd?

   export ROCPROFSYS_TRACE=ON
   export ROCPROFSYS_PROFILE=ON

This multi-layer profiling enables:

* **Understanding communication hierarchies**: See how high-level MPI calls translate to lower-level UCX operations
* **Identifying optimization opportunities**: Detect inefficiencies at different abstraction layers
* **Analyzing GPU-CPU coordination**: Observe interactions between CPU-based MPI communication and GPU-based RCCL collectives
* **Performance debugging**: Trace the full path of data movement from application-level calls to transport-level operations

Best Practices
==============

When profiling communication-intensive applications, consider the following recommendations:

**Start with High-Level Profiling**

* Begin by enabling only MPI or RCCL profiling to understand the overall communication patterns
* Use flat profiles to identify high-overhead communication operations
* Look for functions with high call counts or large cumulative times

**Add Lower-Level Details**

* Enable UCX profiling to understand transport-layer behavior and RDMA utilization
* Use hierarchical profiles to correlate high-level operations with low-level primitives

**Minimize Overhead**

> * Tracing communication operations adds overhead; use sampling mode when precise traces are not required

Reviewer: Do you mean to say that it adds to the size of the output file?

* For large-scale runs, consider enabling profiling on a subset of ranks
* Use ``ROCPROFSYS_SAMPLING_FREQ`` to control sampling rate and balance detail vs. overhead

**Analyze in Context**

* Combine communication profiling with GPU profiling (``ROCPROFSYS_ROCM_DOMAINS``) for heterogeneous applications
* Use ``ROCPROFSYS_TIMEMORY_COMPONENTS`` to add CPU metrics and memory statistics
* Enable process sampling (``ROCPROFSYS_USE_PROCESS_SAMPLING``) for system-level insights

**Leverage Visualization**

> * Use the Perfetto UI to visualize communication timelines and identify bottlenecks

Reviewer: ```suggestion
* Use the Perfetto UI to visualize communication timelines and identify bottlenecks
```
This might also be a good place to mention ROCm Optiq?
